File: Java/Lesson094_FindCriticalAndPseudoCriticalEdgesInMst.java
// =============================================================
// MIT License | @analyticswithharry2026
// GitHub  : https://github.com/analyticswithharry
// YouTube : Analytics with Harry
// =============================================================
// Lesson     : 094 -- Find Critical and Pseudo Critical Edges in MST
// Category   : Advanced Graphs
// Difficulty : Hard
// Study Plan : Day 47
// =============================================================
//
// QUESTION:
//   Given n nodes and weighted edges, find indices of critical and pseudo-critical edges in any MST.
// =============================================================

import java.util.*;
public class Lesson094_FindCriticalAndPseudoCriticalEdgesInMst {
    static class DSU { int[] p,r; int cnt;
        DSU(int n){ p=new int[n]; r=new int[n]; cnt=n; for(int i=0;i<n;i++) p[i]=i; }
        int f(int x){ while(p[x]!=x){p[x]=p[p[x]]; x=p[x];} return x; }
        boolean u(int a,int b){ a=f(a); b=f(b); if(a==b) return false;
            if(r[a]<r[b]){int t=a;a=b;b=t;} p[b]=a; if(r[a]==r[b]) r[a]++; cnt--; return true; }
    }
    int n; int[][] E;
    int kruskal(int skip, int force){
        DSU d=new DSU(n); int w=0;
        if (force>=0){ if(!d.u(E[force][0],E[force][1])) return Integer.MAX_VALUE; w+=E[force][2]; }
        for (int i=0;i<E.length;i++){ if(i==skip) continue; if(d.u(E[i][0],E[i][1])) w+=E[i][2]; }
        return d.cnt==1 ? w : Integer.MAX_VALUE;
    }
    public List<List<Integer>> findCriticalAndPseudoCriticalEdges(int n, int[][] edges){
        this.n=n; int[][] e=new int[edges.length][4];
        for (int i=0;i<edges.length;i++){ e[i][0]=edges[i][0]; e[i][1]=edges[i][1]; e[i][2]=edges[i][2]; e[i][3]=i; }
        Arrays.sort(e,(a,b)->a[2]-b[2]); E=e;
        int base=kruskal(-1,-1);
        List<Integer> crit=new ArrayList<>(), pseu=new ArrayList<>();
        for (int i=0;i<e.length;i++){
            if (kruskal(i,-1) > base) crit.add(e[i][3]);
            else if (kruskal(-1,i) == base) pseu.add(e[i][3]);
        }
        return Arrays.asList(crit, pseu);
    }
    public static void main(String[] a){
        System.out.println(new Lesson094_FindCriticalAndPseudoCriticalEdgesInMst().findCriticalAndPseudoCriticalEdges(5, new int[][]{{0,1,1},{1,2,1},{2,3,2},{0,3,2},{0,4,3},{3,4,3},{1,4,6}}));
    }
}

File: R/Lesson_094_find_critical_and_pseudo_critical_edges_in_mst.R
# =============================================================
# MIT License | @analyticswithharry2026
# GitHub  : https://github.com/analyticswithharry
# YouTube : Analytics with Harry
# =============================================================
# Lesson     : 094 -- Find Critical and Pseudo Critical Edges in MST
# Category   : Advanced Graphs
# Difficulty : Hard
# Study Plan : Day 47
# =============================================================
#
# QUESTION:
#   Given n nodes and weighted edges, find indices of critical and pseudo-critical edges in any MST.
# =============================================================

findCriticalAndPseudoCriticalEdges <- function(n, edges){
  E <- lapply(seq_along(edges), function(i) c(edges[[i]], i-1))
  E <- E[order(sapply(E, function(e) e[3]))]
  dsu_new <- function(n) list(p=0:(n-1), r=rep(0,n), cnt=n)
  dsu_f <- function(d,x){ while(d$p[x+1]!=x){ d$p[x+1] <<- d$p[d$p[x+1]+1]; x <- d$p[x+1] }; x }
  dsu_u <- function(d,a,b){
    a <- dsu_f(d,a); b <- dsu_f(d,b); if(a==b) return(list(d=d, ok=FALSE))
    if(d$r[a+1]<d$r[b+1]){ t<-a; a<-b; b<-t }
    d$p[b+1] <- a; if(d$r[a+1]==d$r[b+1]) d$r[a+1] <- d$r[a+1]+1; d$cnt <- d$cnt-1
    list(d=d, ok=TRUE)
  }
  kruskal <- function(skip, force){
    d <- dsu_new(n); w <- 0
    if (force >= 0){ res <- dsu_u(d, E[[force+1]][1], E[[force+1]][2]); if(!res$ok) return(Inf); d <- res$d; w <- w + E[[force+1]][3] }
    for (i in seq_along(E)){ if (i-1 == skip) next; res <- dsu_u(d, E[[i]][1], E[[i]][2]); if(res$ok){ d <- res$d; w <- w + E[[i]][3] } }
    if (d$cnt != 1) return(Inf); w
  }
  base <- kruskal(-1,-1); crit <- c(); pseu <- c()
  for (i in seq_along(E)){
    if (kruskal(i-1,-1) > base) crit <- c(crit, E[[i]][4])
    else if (kruskal(-1,i-1) == base) pseu <- c(pseu, E[[i]][4])
  }
  list(crit, pseu)
}
print(findCriticalAndPseudoCriticalEdges(5, list(c(0,1,1),c(1,2,1),c(2,3,2),c(0,3,2),c(0,4,3),c(3,4,3),c(1,4,6))))
